refactor(W-14142188): addons-v5: Upgrade addons:upgrade

justinwilaby opened the pull request:
## What
This PR moves `addons:upgrade` to the CLI host and refactors it to use TypeScript and Oclif.

## Test instructions
### Prerequisites: 
1. Checkout this branch and then `yarn install && yarn build`
2. Make sure you have an app with an addon that has plans which can be upgraded such as [Redis Cloud](https://elements.heroku.com/addons/rediscloud)

### Happy Path 1:
1. run `.bin/run addons:upgrade --app <your app> <upgraded plan id>`
2. Observe the output should be the same as running the command on `heroku`

### Happy Path 2:
1. run `.bin/run addons:upgrade --app <your app> <addon id> <upgraded plan id>`
2. Observe the output should be the same as running the command on `heroku`

### Happy Path 3:
1. run `.bin/run addons:upgrade <addon id> <upgraded plan id>`
2. Observe the output should be the same as running the command on `heroku`

### Sad Path 1:
1. run `.bin/run addons:upgrade --app <your app> <plan that does not exist>`
2. Observe the output should be the same as running the command on `heroku`

### Sad Path 2:
1. run `.bin/run addons:upgrade --app <your app>` <--- do not specify a plan
2. Observe the output should be the same as running the command on `heroku`

[W-14142188](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00001al9RCYAY/view)

eablack reviewed (commented):
Compared with the v5 version I notice the `EXAMPLES` section of `help` is missing, and the content of `DESCRIPTION` is missing. We should probably add those in.

Also, we only seem to be supporting one version of how the command works
This one works:

<img width="958" alt="Screenshot 2024-02-21 at 2 06 43 PM" src="https://github.com/heroku/cli/assets/879929/321027be-3ef8-4aaf-8c89-14edf6b5d414">

But this way does not work:

<img width="813" alt="Screenshot 2024-02-21 at 1 52 52 PM" src="https://github.com/heroku/cli/assets/879929/7dfa262a-a1a4-4bf5-a964-922fbbc0bd51">

If we're deciding to remove the second way of upgrading with this command, we should be explicit about it and error when someone tries to do it that way. Right now it kind of looks like it's upgrading successfully but it's not.

justinwilaby commented on packages/cli/src/commands/addons/destroy.ts:
```diff
@@ -32,12 +32,12 @@ export default class Destroy extends Command {
     const force = flags.force || process.env.HEROKU_FORCE === '1'
 
-    const addons = await Promise.all(argv.map(name => resolveAddon(this.heroku, app, name)))
+    const addons = await Promise.all(argv.map(name => resolveAddon(this.heroku, app, name as string)))
```
This resulted from going from untyped to typed for the `resolveAddon()` function. No functional changes, just proper types with these files.

justinwilaby commented:
> Compared with the v5 version I notice the `EXAMPLES` section of `help` is missing, and the content of `DESCRIPTION` is missing. We should probably add those in.
> 
> Also, we only seem to be supporting one version of how the command works This one works:
> 
> <img alt="Screenshot 2024-02-21 at 2 06 43 PM" width="958" src="https://private-user-images.githubusercontent.com/879929/306789350-321027be-3ef8-4aaf-8c89-14edf6b5d414.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1NTc1MzMsIm5iZiI6MTcwODU1NzIzMywicGF0aCI6Ii84Nzk5MjkvMzA2Nzg5MzUwLTMyMTAyN2JlLTNlZjgtNGFhZi04Yzg5LTE0ZWRmNmI1ZDQxNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyMVQyMzEzNTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zYjUxNTFkYTY0NzZmZmVmOTUyODA0ODk3NDcyN2YyOGY3MDIyZGJiYzc2YjEyYmFiMmE5MTY4MzE3NmVlNjQ1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.N5YzmRGLNTdM8Qu72ulXyeawOkh9iU93eyKF_TUkKSw">
> But this way does not work:
> 
> <img alt="Screenshot 2024-02-21 at 1 52 52 PM" width="813" src="https://private-user-images.githubusercontent.com/879929/306788501-7dfa262a-a1a4-4bf5-a964-922fbbc0bd51.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1NTc1MzMsIm5iZiI6MTcwODU1NzIzMywicGF0aCI6Ii84Nzk5MjkvMzA2Nzg4NTAxLTdkZmEyNjJhLWExYTQtNGJmNS1hOTY0LTkyMmZiYmMwYmQ1MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyMVQyMzEzNTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yZDI3NTZiYTk1YmY3OWMzYTc4YTI1YTA5ZmVhYmQzYWZjMjEzMTUwMzllYTlkODdhNmUyYjNmNGU1YzRiYWE1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.xHSMgf7g26fUrn730f2RbwNMUD7HfTn_npS1ZDG4Dbk">
> If we're deciding to remove the second way of upgrading with this command, we should be explicit about it and error when someone tries to do it that way. Right now it kind of looks like it's upgrading successfully but it's not.

The v5 command had top level vars that were reassigned. I cleaned this up but it looks like I missed one deep down in there. Found and fixed. It should work like v5 now.

eablack commented on packages/cli/src/commands/addons/upgrade.ts:
```diff
@@ -0,0 +1,151 @@
+  public async run(): Promise<void> {
+    const ctx = await this.parsed
+    const {flags: {app}, args} = ctx
+    // called with just one argument in the form of `heroku addons:upgrade heroku-redis:hobby`
```
This comment is also on line 88.. was this an unintentional copypaste?

justinwilaby replied:
Removed.

eablack commented on packages/cli/src/lib/addons/resolve.ts:
```diff
@@ -113,7 +116,7 @@ export const attachmentResolver = async (heroku: APIClient, app: string | undefi
       return attachment
     }
-  } catch {}
+  } catch  {}
```
Probably should remove this unintentional whitespace addition

justinwilaby replied:
Applied formatting to the entire file. Done.

eablack reviewed (commented):
LGTM now save for a few minor fix suggestions. I saw your comment about `resolveAddon` .. and yeah, looking at the lib, it looks like a lot has changed with this PR there. It does look more concise the way you've written it, though tbh I can't honestly say I can reason out all the possible ways this might behave differently now.. Are we feeling confident these are safe changes? Should we consider a separate PR for the addon library, or were these changes necessary for the upgrade command conversion?

justinwilaby commented:
> LGTM now save for a few minor fix suggestions. I saw your comment about `resolveAddon` .. and yeah, looking at the lib, it looks like a lot has changed with this PR there. It does look more concise the way you've written it, though tbh I can't honestly say I can reason out all the possible ways this might behave differently now.. Are we feeling confident these are safe changes? Should we consider a separate PR for the addon library, or were these changes necessary for the upgrade command conversion?

Ryan and I had a conversation about this function when he first brought it over to the CLI sources. It wasn't typed properly and was causing issues with ambiguity. I encountered it again with this migration and decided to address it as part of this PR since it was quite trivial to fix and would reduce technical debt. 

Because of the simplicity of this function (a selective memoization strategy), and the unit test coverage, I am 100% confident that the original functionality has been preserved, sans the ambiguity.
